File: April 2025/26.cpp
#include <queue>

class Solution {
  public:
    bool isComplete(Node* root) {
        if (!root) return true;
        
        std::queue<Node*> q;
        q.push(root);
        bool encounteredNull = false;
        
        while (!q.empty()) {
            Node* node = q.front();
            q.pop();
            
            if (!node) {
                encounteredNull = true;
            } else {
                if (encounteredNull) return false; // If we encounter a node after null, it's not complete
                q.push(node->left);
                q.push(node->right);
            }
        }
        
        return true;
    }

    bool isMaxHeap(Node* root) {
        if (!root) return true;

        if (root->left && root->left->data > root->data) return false;
        if (root->right && root->right->data > root->data) return false;
        
        return isMaxHeap(root->left) && isMaxHeap(root->right);
    }

    bool isHeap(Node* tree) {
        return isComplete(tree) && isMaxHeap(tree);
    }
};

File: April 2025/26.py
class Solution:
    # Your Function Should return True/False
    def isHeap(self, root):
        if not root:
            return True
        
        # Helper function to check completeness
        def isComplete(root):
            queue = [(root, 1)]
            index = 0
            
            while queue:
                node, i = queue.pop(0)
                index += 1
                
                if i != index:
                    return False
                
                if node.left:
                    queue.append((node.left, 2 * i))
                if node.right:
                    queue.append((node.right, 2 * i + 1))
            
            return True
        
        # Helper function to check Max-Heap property
        def isMaxHeap(root):
            if not root:
                return True
            
            if root.left and root.left.data > root.data:
                return False
            if root.right and root.right.data > root.data:
                return False
            
            return isMaxHeap(root.left) and isMaxHeap(root.right)
        
        return isComplete(root) and isMaxHeap(root)
